In [ ]:
from IPython.display import HTML, display
display(HTML('''
<div style="text-align:right;padding:8px 0;">
<button id="toggle-code-btn" style="
  padding:8px 18px;font-size:14px;cursor:pointer;
  background:#4a90d9;color:white;border:none;border-radius:5px;
  box-shadow:0 2px 4px rgba(0,0,0,0.2);
" onclick="
  var cells = document.querySelectorAll('.jp-Cell-inputWrapper');
  var btn = document.getElementById('toggle-code-btn');
  var show = btn.dataset.show !== 'true';
  btn.dataset.show = show;
  btn.textContent = show ? '隐藏代码 [Hide]' : '显示代码 [Show]';
  cells.forEach(function(c){ c.style.display = show ? '' : 'none'; });
">显示代码 [Show]</button>
</div>
'''))


# 1.1 Data Representation | 数据表示

**Cambridge International AS & A Level Computer Science (9618)**

---

## 本节学习目标 Learning Objectives

- 理解二进制、十进制、十六进制数制及其转换
- 掌握二进制加法和减法（补码）
- 理解十六进制的用途
- 理解 BCD（二进制编码十进制）
- 理解 ASCII 和 Unicode 字符集
- 理解存储单位（KB, MB, GB, KiB, MiB, GiB…）

---
## 1. 数制基础 | Number Systems

### 1.1 十进制 (Denary / Decimal) - Base 10

我们日常使用的数字系统，使用数字 **0-9**。

每一位有不同的 **权重 (weight)**：

| 10000 (10⁴) | 1000 (10³) | 100 (10²) | 10 (10¹) | 1 (10⁰) |
|:---:|:---:|:---:|:---:|:---:|
| 3 | 1 | 4 | 2 | 1 |

这个数就是: 3×10000 + 1×1000 + 4×100 + 2×10 + 1×1 = **31421**

### 1.2 二进制 (Binary) - Base 2

计算机内部只有两种状态：**开 (ON=1)** 和 **关 (OFF=0)**。

因此计算机使用二进制数制，只有 **0** 和 **1** 两个数字。

每一位叫做一个 **bit（比特）**，8个 bit = 1 **byte（字节）**。

8位二进制的权重：

| 128 (2⁷) | 64 (2⁶) | 32 (2⁵) | 16 (2⁴) | 8 (2³) | 4 (2²) | 2 (2¹) | 1 (2⁰) |
|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|

例如：`1 1 1 0 1 1 1 0` = 128 + 64 + 32 + 8 + 4 + 2 = **238**

00011001 = 2 + 8 + 16 = 26

---
## 2. 进制转换 | Number Conversion

### 2.1 二进制 → 十进制

方法：哪一位是 1，就把对应的权重加起来。

**例子：** `0 1 1 0 1 0 1 1` = 0 + 64 + 32 + 0 + 8 + 0 + 2 + 1 = **107**

In [ ]:
# 二进制 -> 十进制转换器
def binary_to_denary(binary_str):
    """将二进制字符串转换为十进制数"""
    binary_str = binary_str.replace(' ', '')  # 去掉空格
    weights = [2**i for i in range(len(binary_str)-1, -1, -1)]
    
    total = 0
    print(f"\u4e8c\u8fdb\u5236\u6570: {binary_str}")
    print("\u8ba1\u7b97\u8fc7\u7a0b:")
    parts = []
    for i, (bit, weight) in enumerate(zip(binary_str, weights)):
        if bit == '1':
            total += weight
            parts.append(str(weight))
            print(f"  \u7b2c{i+1}\u4f4d: 1 \u00d7 {weight} = {weight}")
        else:
            print(f"  \u7b2c{i+1}\u4f4d: 0 \u00d7 {weight} = 0")
    print(f"\n\u7ed3\u679c: {' + '.join(parts)} = {total}")
    return total

# 试一试！
binary_to_denary('1 1 1 0 1 1 1 0')

### 2.2 十进制 → 二进制

#### 方法一：填权重法

从最大的权重开始，如果还有剩余就填 1，否则填 0。

**例子：** 十进制 107 → 二进制

| 128 | 64 | 32 | 16 | 8 | 4 | 2 | 1 |
|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|
| 0 | 1 | 1 | 0 | 1 | 0 | 1 | 1 |

---

| -128 | 64 | 32 | 16 | 8 | 4 | 2 | 1 |
|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|


- 107 < 128 → 填 0
- 107 >= 64 → 填 1，剩余 107-64 = 43
- 43 >= 32 → 填 1，剩余 43-32 = 11
- 11 < 16 → 填 0
- 11 >= 8 → 填 1，剩余 11-8 = 3
- 3 < 4 → 填 0
- 3 >= 2 → 填 1，剩余 3-2 = 1
- 1 >= 1 → 填 1，剩余 0

结果：**0 1 1 0 1 0 1 1**

#### 方法二：连续除以 2 法

不断除以 2，记录余数，然后从下往上读余数。

```
107 ÷ 2 = 53 ... 余 1  ↑
 53 ÷ 2 = 26 ... 余 1  │
 26 ÷ 2 = 13 ... 余 0  │ 从下往上读
 13 ÷ 2 =  6 ... 余 1  │
  6 ÷ 2 =  3 ... 余 0  │
  3 ÷ 2 =  1 ... 余 1  │
  1 ÷ 2 =  0 ... 余 1  │
                 余 0  │
```
结果：**0 1 1 0 1 0 1 1**

In [ ]:
# 十进制 -> 二进制转换器（连续除以 2 法）
def denary_to_binary(number, bits=8):
    """将十进制数转换为二进制（连续除以2法）"""
    if number < 0:
        print("\u8d1f\u6570\u8bf7\u4f7f\u7528\u8865\u7801\u8f6c\u6362\uff0c\u8fd9\u91cc\u53ea\u5904\u7406\u6b63\u6570")
        return
    
    original = number
    remainders = []
    
    print(f"\u5c06\u5341\u8fdb\u5236 {number} \u8f6c\u6362\u4e3a\u4e8c\u8fdb\u5236\uff1a")
    print("\u8fde\u7eed\u9664\u4ee5 2\uff1a")
    
    while number > 0:
        remainder = number % 2
        quotient = number // 2
        print(f"  {number} \u00f7 2 = {quotient} ... \u4f59 {remainder}")
        remainders.append(str(remainder))
        number = quotient
    
    # 补零到指定位数
    while len(remainders) < bits:
        remainders.append('0')
    
    binary = ''.join(reversed(remainders))
    print(f"\n\u4ece\u4e0b\u5f80\u4e0a\u8bfb\u4f59\u6570: {' '.join(binary)}")
    print(f"\u7ed3\u679c: {original} (\u5341\u8fdb\u5236) = {binary} (\u4e8c\u8fdb\u5236)")
    return binary

# 试一试！
denary_to_binary(107)

### ✅ 练习 1：进制转换

**请在下方代码框中验证你的答案！**

1. 将以下二进制数转换为十进制：
   - a) `0 0 1 1 0 0 1 1`
   - b) `0 1 1 1 1 1 1 1`
   - c) `0 1 1 1 0 1 0 0`
   - d) `0 0 0 0 1 1 1 1`

2. 将以下十进制数转换为 8 位二进制：
   - a) 41
   - b) 67
   - c) 100
   - d) 255

In [1]:
# 练习 1 - 在这里先手动计算，然后运行代码验证

# 例如：你觉得 00110011 = ?
# 先写下你的答案，然后运行验证

print("=== 1. \u4e8c\u8fdb\u5236 -> \u5341\u8fdb\u5236 ===")
print("a) 00110011 =", int('00110011', 2))
print("b) 01111111 =", int('01111111', 2))
print("c) 01110100 =", int('01110100', 2))
print("d) 00001111 =", int('00001111', 2))

print("\n=== 2. \u5341\u8fdb\u5236 -> \u4e8c\u8fdb\u5236 ===")
for num in [41, 67, 100, 255]:
    print(f"{num} = {bin(num)[2:].zfill(8)}")

=== 1. 二进制 -> 十进制 ===
a) 00110011 = 51
b) 01111111 = 127
c) 01110100 = 116
d) 00001111 = 15

=== 2. 十进制 -> 二进制 ===
41 = 00101001
67 = 01000011
100 = 01100100
255 = 11111111


---
## 3. 二进制加法和减法 | Binary Addition & Subtraction

### 3.1 补码表示法 (Two's Complement)

为了表示负数，我们使用 **二进制补码 (Two's Complement)**。

8位补码的权重变为：

| -128 | 64 | 32 | 16 | 8 | 4 | 2 | 1 |
|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|

最高位（最左边）是 **符号位**：
- 0 开头 = 正数（范围: 0 到 +127）
- 1 开头 = 负数（范围: -128 到 -1）

### 如何求一个数的负数（取补码）？

**两步：**
1. **取反** (Invert): 所有的 0 变 1，1 变 0
2. **加 1**: 在最低位加 1

**例子：** 求 -90

```
+90 = 0 1 0 1 1 0 1 0
取反:  1 0 1 0 0 1 0 1
加 1:              + 1
结果:  1 0 1 0 0 1 1 0 = -90
```

验证: -128 + 32 + 4 + 2 = -90 ✅

In [2]:
# 补码转换演示
def twos_complement(binary_str):
    """对一个二进制数取补码（正变负，负变正）"""
    binary_str = binary_str.replace(' ', '')
    print(f"\u539f\u59cb:   {' '.join(binary_str)}")
    
    # 步骤 1: 取反
    inverted = ''.join('1' if b == '0' else '0' for b in binary_str)
    print(f"\u53d6\u53cd:   {' '.join(inverted)}")
    
    # 步骤 2: 加 1
    result = bin(int(inverted, 2) + 1)[2:]
    # 只保留原始位数
    result = result[-len(binary_str):].zfill(len(binary_str))
    print(f"\u52a0 1:   {' '.join(result)}")
    
    # 计算十进制值
    if result[0] == '1':  # 负数
        val = int(result, 2) - (1 << len(result))
    else:
        val = int(result, 2)
    print(f"\u5341\u8fdb\u5236: {val}")
    return result

print("\u4f8b\u5b50: \u6c42 +90 \u7684\u8865\u7801 (\u5373 -90)")
twos_complement('01011010')

print("\n\u4f8b\u5b50: \u6c42 +104 \u7684\u8865\u7801 (\u5373 -104)")
twos_complement('01101000')

例子: 求 +90 的补码 (即 -90)
原始:   0 1 0 1 1 0 1 0
取反:   1 0 1 0 0 1 0 1
加 1:   1 0 1 0 0 1 1 0
十进制: -90

例子: 求 +104 的补码 (即 -104)
原始:   0 1 1 0 1 0 0 0
取反:   1 0 0 1 0 1 1 1
加 1:   1 0 0 1 1 0 0 0
十进制: -104


'10011000'

### 3.2 二进制加法 (Binary Addition)

规则很简单，和十进制加法类似：

| 加法 | 结果 |
|:---:|:---:|
| 0 + 0 | 0 |
| 0 + 1 | 1 |
| 1 + 0 | 1 |
| 1 + 1 | 0，进位 1 |
| 1 + 1 + 1 | 1，进位 1 |

**例子 1：** 37 + 58 = 95

```
  0 0 1 0 0 1 0 1   (37)
+ 0 0 1 1 1 0 1 0   (58)
-------------------
  0 1 0 1 1 1 1 1   (95) ✅
```

### 溢出 (Overflow)

**例子 2：** 82 + 69 = 151，但 8 位最大只能表示 +127！

```
  0 1 0 1 0 0 1 0   (82)
+ 0 1 0 0 0 1 0 1   (69)
-------------------
  1 0 0 1 0 1 1 1   (-105 ?!) ❌ 溢出！
```

两个正数相加得到负数 = **溢出 (Overflow)**！结果超出了 8 位能表示的范围。

In [3]:
# 二进制加法演示
def binary_addition(a_str, b_str):
    """演示两个 8 位二进制数的加法"""
    a_str = a_str.replace(' ', '')
    b_str = b_str.replace(' ', '')
    
    a_val = int(a_str, 2)
    b_val = int(b_str, 2)
    # 补码解释
    a_signed = a_val if a_val < 128 else a_val - 256
    b_signed = b_val if b_val < 128 else b_val - 256
    
    result = (a_val + b_val) & 0xFF  # 只保留 8 位
    result_signed = result if result < 128 else result - 256
    
    print(f"  {' '.join(a_str)}   ({a_signed})")
    print(f"+ {' '.join(b_str)}   ({b_signed})")
    print(f"-" * 25)
    print(f"  {' '.join(bin(result)[2:].zfill(8))}   ({result_signed})")
    
    expected = a_signed + b_signed
    if result_signed != expected:
        print(f"\n\u26a0\ufe0f \u6ea2\u51fa\uff01\u6b63\u786e\u7b54\u6848\u5e94\u8be5\u662f {expected}\uff0c\u4f46 8\u4f4d\u65e0\u6cd5\u8868\u793a")
    else:
        print(f"\n\u2705 \u7ed3\u679c\u6b63\u786e!")

print("\u4f8b\u5b50 1: 37 + 58")
binary_addition('00100101', '00111010')

print("\n" + "="*30)
print("\n\u4f8b\u5b50 2: 82 + 69 (\u6ea2\u51fa)")
binary_addition('01010010', '01000101')

例子 1: 37 + 58
  0 0 1 0 0 1 0 1   (37)
+ 0 0 1 1 1 0 1 0   (58)
-------------------------
  0 1 0 1 1 1 1 1   (95)

✅ 结果正确!


例子 2: 82 + 69 (溢出)
  0 1 0 1 0 0 1 0   (82)
+ 0 1 0 0 0 1 0 1   (69)
-------------------------
  1 0 0 1 0 1 1 1   (-105)

⚠️ 溢出！正确答案应该是 151，但 8位无法表示


### 3.3 二进制减法 (Binary Subtraction)

减法的关键思路：**A - B = A + (-B)**

即：将被减数取补码，然后做加法。

**例子：** 95 - 68

1. 95 = `0 1 0 1 1 1 1 1`
2. 68 = `0 1 0 0 0 1 0 0` → 取补码得 -68 = `1 0 1 1 1 1 0 0`
3. 加法：
```
  0 1 0 1 1 1 1 1   (+95)
+ 1 0 1 1 1 1 0 0   (-68)
-------------------
1 0 0 0 1 1 0 1 1
```
4. 第 9 位的进位丢弃，结果 = `0 0 0 1 1 0 1 1` = **27** ✅

**例子：** 49 - 80

1. 49 = `0 0 1 1 0 0 0 1`
2. 80 = `0 1 0 1 0 0 0 0` → 取补码得 -80 = `1 0 1 1 0 0 0 0`
3. 加法：
```
  0 0 1 1 0 0 0 1   (+49)
+ 1 0 1 1 0 0 0 0   (-80)
-------------------
  1 1 1 0 0 0 0 1   = -31 ✅
```

In [5]:
# 二进制减法演示
def binary_subtraction(a, b):
    """演示 a - b 的二进制减法过程"""
    print(f"\u8ba1\u7b97: {a} - {b}")
    print()
    
    a_bin = bin(a)[2:].zfill(8)
    b_bin = bin(b)[2:].zfill(8)
    print(f"\u6b65\u9aa4 1: \u8f6c\u6362\u4e3a\u4e8c\u8fdb\u5236")
    print(f"  {a} = {' '.join(a_bin)}")
    print(f"  {b} = {' '.join(b_bin)}")
    
    # 取补码
    neg_b = ((~int(b_bin, 2)) + 1) & 0xFF
    neg_b_bin = bin(neg_b)[2:].zfill(8)
    print(f"\n\u6b65\u9aa4 2: \u5bf9 {b} \u53d6\u8865\u7801")
    inv = ''.join('1' if c == '0' else '0' for c in b_bin)
    print(f"  \u53d6\u53cd: {' '.join(inv)}")
    print(f"  \u52a0 1:  {' '.join(neg_b_bin)} = -{b}")
    
    # 相加
    result = (int(a_bin, 2) + neg_b) & 0xFF
    result_bin = bin(result)[2:].zfill(8)
    result_val = result if result < 128 else result - 256
    
    print(f"\n\u6b65\u9aa4 3: \u76f8\u52a0")
    print(f"  {' '.join(a_bin)}   (+{a})")
    print(f"+ {' '.join(neg_b_bin)}   (-{b})")
    print(f"-" * 25)
    print(f"  {' '.join(result_bin)}   ({result_val})")
    print(f"\n\u7ed3\u679c: {a} - {b} = {result_val} \u2705")

binary_subtraction(95, 68)
print("\n" + "="*35 + "\n")
binary_subtraction(49, 80)

计算: 95 - 68

步骤 1: 转换为二进制
  95 = 0 1 0 1 1 1 1 1
  68 = 0 1 0 0 0 1 0 0

步骤 2: 对 68 取补码
  取反: 1 0 1 1 1 0 1 1
  加 1:  1 0 1 1 1 1 0 0 = -68

步骤 3: 相加
  0 1 0 1 1 1 1 1   (+95)
+ 1 0 1 1 1 1 0 0   (-68)
-------------------------
  0 0 0 1 1 0 1 1   (27)

结果: 95 - 68 = 27 ✅


计算: 49 - 80

步骤 1: 转换为二进制
  49 = 0 0 1 1 0 0 0 1
  80 = 0 1 0 1 0 0 0 0

步骤 2: 对 80 取补码
  取反: 1 0 1 0 1 1 1 1
  加 1:  1 0 1 1 0 0 0 0 = -80

步骤 3: 相加
  0 0 1 1 0 0 0 1   (+49)
+ 1 0 1 1 0 0 0 0   (-80)
-------------------------
  1 1 1 0 0 0 0 1   (-31)

结果: 49 - 80 = -31 ✅


### ✅ 练习 2：二进制加减法

请手动计算以下题目，然后用代码验证：

1. `0 0 1 1 1 0 0 1` + `0 0 1 0 1 0 0 1` = ?
2. `0 1 0 0 1 0 1 1` + `0 0 1 0 0 0 1 1` = ?
3. `0 1 1 0 0 0 1 1` - `0 0 1 1 0 0 0 0` = ?
4. `0 1 1 1 1 1 1 1` - `0 1 0 1 1 0 1 0` = ?
5. 将 +114 转换为 8 位二进制
6. 将 -14 转换为 8 位二进制（补码）



2 补码 -> 先取反 再加1

\n

1 补码 -> 取反



In [8]:
# 练习 2 - 验证答案



print("\n4. 01111111 - 01011010")
binary_subtraction(0b01111111, 0b01011010)




4. 01111111 - 01011010
计算: 127 - 90

步骤 1: 转换为二进制
  127 = 0 1 1 1 1 1 1 1
  90 = 0 1 0 1 1 0 1 0

步骤 2: 对 90 取补码
  取反: 1 0 1 0 0 1 0 1
  加 1:  1 0 1 0 0 1 1 0 = -90

步骤 3: 相加
  0 1 1 1 1 1 1 1   (+127)
+ 1 0 1 0 0 1 1 0   (-90)
-------------------------
  0 0 1 0 0 1 0 1   (37)

结果: 127 - 90 = 37 ✅


---
## 4. 存储单位 | Memory Units

### SI 十进制前缀 (Decimal Prefixes)

| 名称 | 符号 | 字节数 |
|:---|:---:|---:|
| 1 Kilobyte | 1 KB | 1,000 |
| 1 Megabyte | 1 MB | 1,000,000 |
| 1 Gigabyte | 1 GB | 1,000,000,000 |
| 1 Terabyte | 1 TB | 1,000,000,000,000 |
| 1 Petabyte | 1 PB | 1,000,000,000,000,000 |

### IEC 二进制前缀 (Binary Prefixes)

因为计算机内存基于 2 的幂次，IEC 提出了更精确的单位：

| 名称 | 符号 | 字节数 |
|:---|:---:|---:|
| 1 Kibibyte | 1 KiB | 2¹⁰ = 1,024 |
| 1 Mebibyte | 1 MiB | 2²⁰ = 1,048,576 |
| 1 Gibibyte | 1 GiB | 2³⁰ = 1,073,741,824 |
| 1 Tebibyte | 1 TiB | 2⁴⁰ = 1,099,511,627,776 |

**重要区别：**
- 硬盘厂商通常用 SI 前缀（看起来更大）
- 内存 (RAM) 应该用 IEC 前缀（更准确）
- 这就是为什么你买的 "1 TB" 硬盘实际显示的容量比 1 TB 少一些！

In [9]:
# 存储单位对比
print(f"{'\u540d\u79f0':<15} {'SI (\u5341\u8fdb\u5236)':>20} {'IEC (\u4e8c\u8fdb\u5236)':>20} {'\u5dee\u5f02':>10}")
print("-" * 70)

units = [
    ('Kilo/Kibi', 10**3, 2**10),
    ('Mega/Mebi', 10**6, 2**20),
    ('Giga/Gibi', 10**9, 2**30),
    ('Tera/Tebi', 10**12, 2**40),
]

for name, si, iec in units:
    diff_pct = (iec - si) / si * 100
    print(f"{name:<15} {si:>20,} {iec:>20,} {diff_pct:>9.1f}%")

print("\n\u4f8b\u5b50: 64 GiB RAM \u80fd\u5b58\u50a8\u591a\u5c11\u5b57\u8282?")
print(f"  64 \u00d7 2\u00b3\u2070 = {64 * 2**30:,} bytes")

名称                          SI (十进制)            IEC (二进制)         差异
----------------------------------------------------------------------
Kilo/Kibi                      1,000                1,024       2.4%
Mega/Mebi                  1,000,000            1,048,576       4.9%
Giga/Gibi              1,000,000,000        1,073,741,824       7.4%
Tera/Tebi          1,000,000,000,000    1,099,511,627,776      10.0%

例子: 64 GiB RAM 能存储多少字节?
  64 × 2³⁰ = 68,719,476,736 bytes


---
## 5. 十六进制 | Hexadecimal (Base 16)

### 5.1 基本概念

十六进制使用 16 个符号：**0-9** 和 **A-F**

| 十进制 | 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12 | 13 | 14 | 15 |
|:---|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|
| 十六进制 | 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | A | B | C | D | E | F |
| 二进制 | 0000 | 0001 | 0010 | 0011 | 0100 | 0101 | 0110 | 0111 | 1000 | 1001 | 1010 | 1011 | 1100 | 1101 | 1110 | 1111 |

因为 16 = 2⁴，所以 **1 个十六进制位 = 4 个二进制位**

### 5.2 二进制 ↔ 十六进制转换

**二进制 → 十六进制：** 从右往左每 4 位一组，每组转换为一个十六进制位

```
例子: 1011 1110 0001
         B    E    1
结果: BE1
```

**十六进制 → 二进制：** 每个十六进制位展开为 4 位二进制

```
例子: 4 5 A
      0100 0101 1010
结果: 010001011010
```

### 5.3 为什么用十六进制？

1. **内存转储 (Memory Dump)：** 调试时看内存内容，十六进制比二进制短得多
   - 二进制: `10110101101001000001101011111100`
   - 十六进制: `B5A41AFC` ← 容易看得多！

2. **颜色代码：** 网页颜色如 `#FF5733` 就是十六进制

3. **MAC 地址：** 如 `00:1A:2B:3C:4D:5E`




In [10]:
# 二进制 <-> 十六进制转换演示

def binary_to_hex(binary_str):
    """二进制转十六进制，展示过程"""
    binary_str = binary_str.replace(' ', '')
    # 补零使长度是4的倍数
    while len(binary_str) % 4 != 0:
        binary_str = '0' + binary_str
    
    print(f"\u4e8c\u8fdb\u5236: {binary_str}")
    groups = [binary_str[i:i+4] for i in range(0, len(binary_str), 4)]
    hex_digits = []
    print("\u5206\u7ec4:")
    for g in groups:
        h = hex(int(g, 2))[2:].upper()
        hex_digits.append(h)
        print(f"  {g} = {h}")
    result = ''.join(hex_digits)
    print(f"\u7ed3\u679c: {result}")
    return result

def hex_to_binary(hex_str):
    """十六进制转二进制，展示过程"""
    hex_str = hex_str.replace(' ', '').upper()
    print(f"\u5341\u516d\u8fdb\u5236: {hex_str}")
    print("\u5c55\u5f00:")
    binary_parts = []
    for h in hex_str:
        b = bin(int(h, 16))[2:].zfill(4)
        binary_parts.append(b)
        print(f"  {h} = {b}")
    result = ''.join(binary_parts)
    print(f"\u7ed3\u679c: {result}")
    return result

print("=== \u4e8c\u8fdb\u5236 -> \u5341\u516d\u8fdb\u5236 ===")
binary_to_hex('10111110 0001')

print("\n=== \u5341\u516d\u8fdb\u5236 -> \u4e8c\u8fdb\u5236 ===")
hex_to_binary('4 5 A')

=== 二进制 -> 十六进制 ===
二进制: 101111100001
分组:
  1011 = B
  1110 = E
  0001 = 1
结果: BE1

=== 十六进制 -> 二进制 ===
十六进制: 45A
展开:
  4 = 0100
  5 = 0101
  A = 1010
结果: 010001011010


'010001011010'

### ✅ 练习 3：十六进制转换

**将以下二进制数转换为十六进制：**
1. `1 1 0 0 0 0 1 1`
2. `1 1 1 1 0 1 1 1`
3. `1 0 0 1 1 1 1 1 1 1`

**将以下十六进制数转换为二进制：**
4. `6C`
5. `A00`
6. `DA47`

In [11]:
# 练习 3 - 验证答案

print("=== \u4e8c\u8fdb\u5236 -> \u5341\u516d\u8fdb\u5236 ===")
for i, b in enumerate(['11000011', '11110111', '1001111111'], 1):
    print(f"\n{i}.")
    binary_to_hex(b)

print("\n\n=== \u5341\u516d\u8fdb\u5236 -> \u4e8c\u8fdb\u5236 ===")
for i, h in enumerate(['6C', 'A00', 'DA47'], 4):
    print(f"\n{i}.")
    hex_to_binary(h)

=== 二进制 -> 十六进制 ===

1.
二进制: 11000011
分组:
  1100 = C
  0011 = 3
结果: C3

2.
二进制: 11110111
分组:
  1111 = F
  0111 = 7
结果: F7

3.
二进制: 001001111111
分组:
  0010 = 2
  0111 = 7
  1111 = F
结果: 27F


=== 十六进制 -> 二进制 ===

4.
十六进制: 6C
展开:
  6 = 0110
  C = 1100
结果: 01101100

5.
十六进制: A00
展开:
  A = 1010
  0 = 0000
  0 = 0000
结果: 101000000000

6.
十六进制: DA47
展开:
  D = 1101
  A = 1010
  4 = 0100
  7 = 0111
结果: 1101101001000111


---
## 6. BCD（二进制编码十进制）| Binary-Coded Decimal

### 6.1 基本概念

BCD 用 **4 位二进制** 表示 **每一个十进制数字**（而不是整个数）。

```
0 = 0000    5 = 0101
1 = 0001    6 = 0110
2 = 0010    7 = 0111
3 = 0011    8 = 1000
4 = 0100    9 = 1001
```

**例子：** 十进制 3165 的 BCD 表示

```
3    1    6    5
0011 0001 0110 0101
```

### 6.2 BCD vs 普通二进制的区别

| | 普通二进制 | BCD |
|:---|:---|:---|
| 数字 42 | 00101010 (8位) | 0100 0010 (8位) |
| 转换方式 | 整个数转换 | 每个数字单独转换 |

### 6.3 BCD 的用途

- **计算器/时钟显示：** 每个数字直接对应 BCD，容易显示
- **金融计算：** 避免二进制小数的精度问题，精确表示货币值

In [ ]:
# BCD 转换演示

def denary_to_bcd(number):
    """十进制转 BCD"""
    num_str = str(number)
    print(f"\u5341\u8fdb\u5236 {number} \u8f6c BCD:")
    bcd_parts = []
    for digit in num_str:
        bcd = bin(int(digit))[2:].zfill(4)
        bcd_parts.append(bcd)
        print(f"  {digit} -> {bcd}")
    result = ' '.join(bcd_parts)
    print(f"\u7ed3\u679c: {result}")
    return result

def bcd_to_denary(bcd_str):
    """将 BCD 转换为十进制"""
    bcd_str = bcd_str.replace(' ', '')
    print(f"BCD {bcd_str} \u8f6c\u5341\u8fdb\u5236:")
    groups = [bcd_str[i:i+4] for i in range(0, len(bcd_str), 4)]
    digits = []
    for g in groups:
        d = int(g, 2)
        digits.append(str(d))
        print(f"  {g} -> {d}")
    result = ''.join(digits)
    print(f"\u7ed3\u679c: {result}")
    return int(result)

denary_to_bcd(3165)
print()
bcd_to_denary('1001 0011 0111')

print("\n--- BCD vs \u666e\u901a\u4e8c\u8fdb\u5236\u5bf9\u6bd4 ---")
num = 42
print(f"\u6570\u5b57 {num}:")
print(f"  \u666e\u901a\u4e8c\u8fdb\u5236: {bin(num)[2:].zfill(8)}")
denary_to_bcd(num)

### ✅ 练习 4：BCD 转换

1. 将以下十进制数转换为 BCD：
   - a) 271
   - b) 5006
   - c) 7990

2. 将以下 BCD 转换为十进制：
   - a) `1001 0011 0111`
   - b) `0111 0111 0110 0010`

In [ ]:
# 练习 4 - 验证答案

print("=== \u5341\u8fdb\u5236 -> BCD ===")
for num in [271, 5006, 7990]:
    denary_to_bcd(num)
    print()

print("=== BCD -> \u5341\u8fdb\u5236 ===")
bcd_to_denary('1001 0011 0111')
print()
bcd_to_denary('0111 0111 0110 0010')

---
## 7. 字符集 | Character Sets

### 7.1 ASCII

**ASCII (American Standard Code for Information Interchange)**

- 用 **7 位二进制** 表示字符，可以表示 **128** 个字符 (0-127)
- 包括：大小写字母 (A-Z, a-z)、数字 (0-9)、标点符号、控制字符
- 存储时通常用 1 个字节 (8 位)

**常见 ASCII 值：**

| 字符 | 十进制 | 十六进制 |
|:---:|:---:|:---:|
| 空格 | 32 | 20 |
| '0' | 48 | 30 |
| 'A' | 65 | 41 |
| 'a' | 97 | 61 |

**注意：** 大写字母和小写字母的 ASCII 值相差 **32**！

### 7.2 Unicode

ASCII 只能表示英文字符，无法表示中文、日文等。

**Unicode** 可以表示世界上所有语言的字符：
- 前 128 个字符与 ASCII 相同
- 使用 **2-4 个字节** 表示一个字符
- 能表示超过 **100,000** 个字符
- 包括中文、日文、韩文、阿拉伯文、Emoji 等

**缺点：** 占用更多存储空间

In [ ]:
# ASCII 码探索

print("=== \u5e38见 ASCII \u7801\u8868 ===")
print(f"{'\u5b57\u7b26':<6} {'\u5341\u8fdb\u5236':<8} {'\u5341\u516d\u8fdb\u5236':<8} {'\u4e8c\u8fdb\u5236'}")
print("-" * 35)

chars = [' ', '0', '9', 'A', 'Z', 'a', 'z']
for c in chars:
    code = ord(c)
    name = 'SPACE' if c == ' ' else c
    print(f"{name:<6} {code:<8} {hex(code):<8} {bin(code)[2:].zfill(8)}")

print("\n=== \u5927\u5c0f\u5199\u5b57\u6bcd\u7684\u5173\u7cfb ===")
print(f"'A' = {ord('A')},  'a' = {ord('a')},  \u5dee\u503c = {ord('a') - ord('A')}")
print(f"'B' = {ord('B')},  'b' = {ord('b')},  \u5dee\u503c = {ord('b') - ord('B')}")
print("\u6240\u4ee5\u5927\u5199\u8f6c\u5c0f\u5199\u53ea\u9700 +32\uff0c\u5c0f\u5199\u8f6c\u5927\u5199\u53ea\u9700 -32")

print("\n=== Unicode \u793a\u4f8b ===")
examples = [('\u4f60', '\u4f60'), ('\u597d', '\u597d'), ('\u2764', '\u2764'), ('\u2705', '\u2705')]
for name, char in examples:
    code = ord(char)
    print(f"'{char}' -> U+{code:04X} (\u5341\u8fdb\u5236 {code}), \u9700\u8981 {len(char.encode('utf-8'))} \u5b57\u8282(UTF-8)")

### ✅ 练习 5：字符集

1. 字符 'H' 的 ASCII 码是多少？用二进制和十六进制表示。
2. 如果一个字符的 ASCII 码是 01000100（二进制），这个字符是什么？
3. 字符串 "Hi" 用 ASCII 存储需要多少字节？用 Unicode (UTF-8) 呢？
4. 为什么 ASCII 不能表示中文？

In [ ]:
# 练习 5 - 验证答案

print("1. \u5b57\u7b26 'H' \u7684 ASCII \u7801:")
h_code = ord('H')
print(f"   \u5341\u8fdb\u5236: {h_code}")
print(f"   \u4e8c\u8fdb\u5236: {bin(h_code)[2:].zfill(8)}")
print(f"   \u5341\u516d\u8fdb\u5236: {hex(h_code)}")

print(f"\n2. \u4e8c\u8fdb\u5236 01000100 = \u5341\u8fdb\u5236 {int('01000100', 2)} = \u5b57\u7b26 '{chr(int('01000100', 2))}'")

print(f"\n3. 'Hi' \u7684\u5b58\u50a8:")
print(f"   ASCII: {len('Hi')} \u5b57\u8282 (\u6bcf\u4e2a\u5b57\u7b26 1 \u5b57\u8282)")
print(f"   UTF-8: {len('Hi'.encode('utf-8'))} \u5b57\u8282 (\u82f1\u6587\u5b57\u7b26\u5728 UTF-8 \u4e2d\u4e5f\u662f 1 \u5b57\u8282)")

print(f"\n4. ASCII \u53ea\u7528 7 \u4f4d\u4e8c\u8fdb\u5236\uff0c\u6700\u591a\u8868\u793a 128 \u4e2a\u5b57\u7b26")
print(f"   \u4e2d\u6587\u6709\u51e0\u4e07\u4e2a\u5b57\uff0c\u8fdc\u8d85 128\uff0c\u6240\u4ee5\u9700\u8981 Unicode")

---
## 8. 综合练习 | Comprehensive Exercises

### 练习 A：进制转换综合

1. 将十进制 189 分别转换为：
   - a) 8位二进制
   - b) 十六进制
   - c) BCD

2. 将十六进制 `9C` 转换为：
   - a) 二进制
   - b) 十进制

3. 计算 `01100011` + `00110000`，给出二进制和十进制结果

4. 计算 75 - 120，给出 8 位二进制结果

### 练习 B：思考题

5. 为什么计算机使用二进制而不是十进制？

6. 8 位二进制补码能表示的数值范围是什么？

7. 一个字符串 "HELLO" 用 ASCII 存储需要多少字节？如果用 Unicode (UTF-8) 呢？

8. 为什么金融系统使用 BCD 而不是普通二进制？

In [ ]:
# 综合练习答案（先自己做，再运行验证！）

print("=== \u7ec3\u4e60 A ===")
print("\n1a) 189 -> \u4e8c\u8fdb\u5236:", bin(189)[2:].zfill(8))
print("1b) 189 -> \u5341\u516d\u8fdb\u5236:", hex(189)[2:].upper())
print("1c) 189 -> BCD:", end=" ")
denary_to_bcd(189)

print("\n2a) 9C -> \u4e8c\u8fdb\u5236:")
hex_to_binary('9C')
print(f"2b) 9C -> \u5341\u8fdb\u5236: {int('9C', 16)}")

print("\n3) 01100011 + 00110000:")
binary_addition('01100011', '00110000')

print("\n4) 75 - 120:")
binary_subtraction(75, 120)

In [12]:
# 练习 B - 思考题参考答案

print("""
5. \u4e3a\u4ec0\u4e48\u8ba1\u7b97\u673a\u4f7f\u7528\u4e8c\u8fdb\u5236\uff1f
   \u56e0\u4e3a\u8ba1\u7b97\u673a\u5185\u90e8\u7531\u6570\u767e\u4e07\u4e2a\u5fae\u5c0f\u7684\u5f00\u5173\u7ec4\u6210\uff0c
   \u6bcf\u4e2a\u5f00\u5173\u53ea\u6709\u4e24\u79cd\u72b6\u6001\uff1aON(1) \u548c OFF(0)\uff0c
   \u8fd9\u81ea\u7136\u5bf9\u5e94\u4e8c\u8fdb\u5236\u6570\u5236\u3002

6. 8\u4f4d\u4e8c\u8fdb\u5236\u8865\u7801\u7684\u8303\u56f4\uff1a
""")
print(f"   \u6700\u5c0f\u503c: -128  (10000000)")
print(f"   \u6700\u5927\u503c: +127  (01111111)")
print(f"   \u8303\u56f4: -128 \u5230 +127")

print(f"""
7. "HELLO" \u7684\u5b58\u50a8:
   ASCII: {len('HELLO')} \u5b57\u8282 (\u6bcf\u4e2a\u82f1\u6587\u5b57\u7b26 1 \u5b57\u8282)
   UTF-8: {len('HELLO'.encode('utf-8'))} \u5b57\u8282 (\u82f1\u6587\u5b57\u7b26\u5728 UTF-8 \u4e2d\u4e5f\u662f 1 \u5b57\u8282)
   \u6ce8: \u5982\u679c\u662f\u4e2d\u6587\u5b57\u7b26\uff0cUTF-8 \u6bcf\u4e2a\u5b57\u7b26\u9700\u8981 3 \u5b57\u8282

8. \u4e3a\u4ec0\u4e48\u91d1\u878d\u7cfb\u7edf\u7528 BCD\uff1f
   \u56e0\u4e3a\u666e\u901a\u4e8c\u8fdb\u5236\u65e0\u6cd5\u7cbe\u786e\u8868\u793a\u67d0\u4e9b\u5c0f\u6570(\u5982 0.1)\uff0c
   \u8fd9\u4f1a\u5bfc\u81f4\u8d27\u5e01\u8ba1\u7b97\u4ea7\u751f\u7d2f\u79ef\u8bef\u5dee\u3002
   BCD \u6bcf\u4e2a\u5341\u8fdb\u5236\u4f4d\u5355\u72ec\u7f16\u7801\uff0c\u80fd\u7cbe\u786e\u8868\u793a\u8d27\u5e01\u503c\u3002
""")

# \u6f14\u793a\u4e8c\u8fdb\u5236\u5c0f\u6570\u7684\u7cbe\u5ea6\u95ee\u9898
print("--- \u4e8c\u8fdb\u5236\u5c0f\u6570\u7cbe\u5ea6\u95ee\u9898\u6f14\u793a ---")
print(f"0.1 + 0.2 = {0.1 + 0.2}  (\u4e0d\u662f\u7cbe\u786e\u7684 0.3\uff01)")
print("\u8fd9\u5c31\u662f\u4e3a\u4ec0\u4e48\u91d1\u878d\u7cfb\u7edf\u9700\u8981 BCD\uff01")


5. 为什么计算机使用二进制？
   因为计算机内部由数百万个微小的开关组成，
   每个开关只有两种状态：ON(1) 和 OFF(0)，
   这自然对应二进制数制。

6. 8位二进制补码的范围：

   最小值: -128  (10000000)
   最大值: +127  (01111111)
   范围: -128 到 +127

7. "HELLO" 的存储:
   ASCII: 5 字节 (每个英文字符 1 字节)
   UTF-8: 5 字节 (英文字符在 UTF-8 中也是 1 字节)
   注: 如果是中文字符，UTF-8 每个字符需要 3 字节

8. 为什么金融系统用 BCD？
   因为普通二进制无法精确表示某些小数(如 0.1)，
   这会导致货币计算产生累积误差。
   BCD 每个十进制位单独编码，能精确表示货币值。

--- 二进制小数精度问题演示 ---
0.1 + 0.2 = 0.30000000000000004  (不是精确的 0.3！)
这就是为什么金融系统需要 BCD！


---
## 知识点总结 | Summary

| 主题 | 核心要点 |
|:---|:---|
| **二进制** | 基数为2，只有0和1，计算机的基础 |
| **十六进制** | 基数为16，用0-9和A-F，1个hex位=4个binary位 |
| **补码** | 取反+1，用于表示负数，8位范围: -128~+127 |
| **溢出** | 计算结果超出可表示范围 |
| **BCD** | 每个十进制位用4位二进制表示，用于精确表示 |
| **ASCII** | 7位编码，128个字符，只支持英文 |
| **Unicode** | 支持全球所有语言，前128个与ASCII相同 |
| **存储单位** | KB/MB/GB(十进制) vs KiB/MiB/GiB(二进制) |

---

**下一步学习：** 1.2 Multimedia（多媒体）- 图像、声音的数字化表示